README.rst
Questo prodotto si occupa di aggiungere alcune funzionalità a `collective.tablepage`__ per gestire i casi in cui
il prodotto sia stato usato per sopperire alle richieste dell'`art. 1 comma 32 Legge n. 190/2012`__ e che sia quindi
necessario esporre i dati in formato XML.

__ http://plone.org/products/collective.tablepage
__ http://www.avcp.it/portal/public/classic/AttivitaAutorita/AttiDellAutorita/_Atto?ca=5397

.. image:: http://blog.redturtle.it/pypi-images/ospfe.singole_procedure/ospfe.singole_procedure-0.1.0-02.png/image_large
   :alt: Configurazione tabella
   :target: http://blog.redturtle.it/pypi-images/ospfe.singole_procedure/ospfe.singole_procedure-0.1.0-02.png
   :align: center

Una copia delle specifiche tecniche è inclusa nel sorgente del prodotto.

.. contents::

Installazione
=============

Il prodotto non è rilasciato ufficialmente sul `Cheeseshop`__, per poterlo installare da buildout è necessario
modificare la configurazione del medesimo::

    [buildout]
    ...
    find-links =
        ...
        https://github.com/PloneGov-IT/ospfe.singole_procedure/releases/download/x.y.z/ospfe.singole_procedure-x.y.z.zip

... dove *x.y.z* rappresenta la versione del prodotto da utilizzare, ad esempio::

    https://github.com/PloneGov-IT/ospfe.singole_procedure/releases/download/0.3.0/ospfe.singole_procedure-0.3.0.zip

__ http://pypi.python.org/

Come usarlo
===========

Attivazione
-----------

Va scelta una pagina con tabella specifica e va aggiunta un'interfaccia marcatore. Per fare questo va richiamato
l'URL del documento aggiungendovi in fondo "/manage_interfaces". Esempio:

    http://host.com/percorso/alla/pagina-con-tabella/manage_interfaces

L'interfaccia da aggiungere è ``ospfe.singole_procedure.interfaces.ISingoleProcedureInterface``.

Da questo momento in poi questa singola pagina con tabella ha nuove funzionalità sotto descritte.

Nuove colonne
-------------

Il prodotto aggiunge una serie di nuovi tipi di colonne al solo scopo di aiutrare la compilazione della tabella.

``CIG``
    Un campo stringa di non più di 10 caratteri
``Struttura proponente``
    Un campo stringa che al salvataggio verifica i dati inseriti.
    
    Tali dati *devono* includere una denominazione (massimo 250 caratteri) seguita o preceduta dal codice fiscale o
    partita I.V.A.
``Oggetto``
    Oggetto del bando, limitato a 250 caratteri
``Procedura scelta contraente``
    Un campo di selezione che già utilizza il vocabolario definito dallo schema XSD, senza doverlo configurare.
``Operatori / Aggiudicatari``
    Un campo valido per entrambi i tipi di dato. Permette di definire una sequenza (un elemento per riga) di partecipanti
    o aggiundicatari del bando.
        
    Ogni elemento deve essere composto dalla denominazione seguita dal codice fiscale o partita I.V.A.
    
    In presenza di consorziate va infine posto (tra parentesi) il ruolo, secondo il vocabolario definito dallo schema XSD.
    Le corsorziate definiscono quindi un gruppo, che va separato dagli altri gruppi, o dai partecipanti/aggiudicatari
    singoli,
    da una riga vuota.
``Dal-Al``
    Un campo stringa per l'inserimento delle due date. Le date devono essere nel formato AAAA-MM-GG o anche nel
    formato GG-MM-AAAA, e possono essere separate da un carattere separatore quale "-" e "/".

Dato che non si vuole "inquinare" il normale uso della pagina con tabella, l'attivazione delle nuove colonne avviene
in modo puntuale previa attivazione dell'interfaccia, come discusso alla sezione precedente.

Struttura del documento
-----------------------

.. image:: http://blog.redturtle.it/pypi-images/ospfe.singole_procedure/ospfe.singole_procedure-0.1.0-01.png/image_preview
   :alt: Configurazione tabella
   :target: http://blog.redturtle.it/pypi-images/ospfe.singole_procedure/ospfe.singole_procedure-0.1.0-01.png
   :align: right

Il prodotto da' per scontato che la tabella generata abbia una ben definita struttura a *9 colonne*
definite in questo ordine:

* **CIG** - consigliato uso della colonna "CIG" (non obbligatorio, vedere `le FAQ`__)
* **Struttura proponente**  - consigliato uso della colonna "Struttura proponente" e obbligatorietà
* **Oggetto del bando** - consigliato uso della colonna "Oggetto" e obbligatorietà
* **Procedura di scelta del contraente** - consigliato uso della colonna "Procedura scelta contraente"
* **Elenco operatori** - consigliato uso della colonna "Operatori / Aggiudicatari"
* **Aggiudicatario** - consigliato uso della colonna "Operatori / Aggiudicatari"
* **Importo di aggiudicazione** - consigliato uso della colonna "Numerico" o "Monetario"
* **Tempi completamento** - consigliato uso della colonna "Dal-Al"
* **Importo somme liquidate** - consigliato uso della colonna "Numerico" o "Monetario"

__ http://www.avcp.it/portal/public/classic/FAQ/faq_legge190_2012#sezioneC

XML e Validazione
=================

Il file XML richiesto dalla normativa è sempre disponibile richiamando la vista **dataset.xml** sulla pagina con tabella.
Per esempio:

    http://host.com/percorso/alla/pagina-con-tabella/dataset.xml

Prima di esporre il sorgente XML è possibile effettuare una validazione dello stesso. Dalla pagina di modifica della
tabella sarà ora disponibile un nuovo link "*Valida XML*", che eseguirà la validazione secondo lo schema fornito da AVCP.

.. image:: http://blog.redturtle.it/pypi-images/ospfe.singole_procedure/ospfe.singole_procedure-0.1.0-03.png/image_large
   :alt: Configurazione tabella
   :target: http://blog.redturtle.it/pypi-images/ospfe.singole_procedure/ospfe.singole_procedure-0.1.0-03.png
   :align: center

Dato ``annoRiferimento``
------------------------

All'interno del file XML generato è presente un elemento ``annoRiferimento`` che di base viene valorizzato a
*2013*. È possibile cambiare tale scelta aggiungendo al contenuto una CMF property di nome ``anno_riferimento``.

Dipendenze e versioni
=====================

Il prodotto è stato testato su Plone 4.3, ma *dovrebbe* ancora funzionare senza problemi anche su Plone 3.3.

Limiti
======

Non viene fornito un XML per l'indice dei dataset. Ipoteticamente potrebbe essere fatto con una vista apposita
sugli oggetti Collezione (`pull request`__ a riguardo sono ben accette!).

__ https://github.com/PloneGov-IT/ospfe.singole_procedure/pulls

Crediti
=======

Sviluppato col supporto di:

* `Ospedale Sant'Anna, Ferrara`__

  .. image:: http://www.ospfe.it/ospfe-logo.jpg 
     :alt: Logo Ospedale S. Anna
  
* `Azienda USL Ferrara`__

  .. image:: http://www.ausl.fe.it/logo_ausl.gif
     :alt: Logo Azienda AUSL

Questi enti supportano `L'iniziativa PloneGov`__.

__ http://www.ospfe.it/
__ http://www.ausl.fe.it/
__ http://www.plonegov.it/

Autori
======

Questo prodotto è stato sviluppato da RedTurtle Technology.

.. image:: http://www.redturtle.it/redturtle_banner.png
   :alt: RedTurtle Technology Site
   :target: http://www.redturtle.it/


docs/HISTORY.rst
Changelog
=========

0.3.6 (unreleased)
------------------

- Nothing changed yet.


0.3.5 (2020-03-06)
------------------

- Updated definition.
  [daniele]

0.3.4 (2020-01-14)
------------------

- Changed check on effective date. Updated schema as requested.
  [daniele]


0.3.3 (2019-12-11)
------------------

- Updated TypesL190.xsd
  [daniele]

0.3.2 (2019-29-01)
------------------

- Updated TypesL190.xsd
  [daniele]

0.3.1 (2015-01-30)
------------------

- Migliorata affidabilità regexp
  [keul]

0.3.0 (2015-01-15)
------------------

Verifiche di compatibilità con Plone 4.3

- ``annoRiferimento`` viene letto (potenzialmente) da una CMF property.
  Il default rimane al 2013 per compatibilità con l'anno scorso
  [keul]
- Aggiunge funzionalità di debug
  [keul]
- Fix alle regex per gestire casi sfortunati in cui la denominazione può essere
  scambiata per un C.F. o P. IVA.
  [keul]
- Importi nella vista di generazione XML ora sempre in formato stringa
  [keul]
- Rinominata colonna "*Date*" in "*Dal-Al*" per namespace clash con
  nuove colonne di tablepage 0.10
  [keul]

0.2.0 (2014-01-31)
------------------

- Bug: la presenza di costi con valore ``0.00`` non mostrava il nodo
  [keul]
- Permessa la presenza di codice fiscale/P.I. prima della ragione sociale
  (con qualche limite)
  [keul]

0.1.0 (2014-01-29)
------------------

- Initial release


ospfe/singole_procedure/config.py
# -*- coding: utf-8 -*-

STRUTTURA_PROPONENTE_MODEL = r"""^\s*((?P<cf1>(?:[0-9]{11})|(?:[A-Za-z]{6}\d\d[a-zA-Z0-9]{8}))\s+(?P<denominazione1>.+))|((?P<denominazione2>.+)\s+(?P<cf2>(?:[0-9]{11})|(?:[A-Za-z]{6}\d\d[a-zA-Z0-9]{8})))\s*$"""
CF_MODEL = r"""^(
    [A-Za-z]{6}[0-9]{2}[A-Za-z]{1}[0-9]{2}[A-Za-z]{1}[0-9A-Za-z]{3}[A-Za-z]{1} |
    [A-Za-z]{6}[0-9LMNPQRSTUV]{2}[A-Za-z]{1}[0-9LMNPQRSTUV]{2}[A-Za-z]{1}[0-9LMNPQRSTUV]{3}[A-Za-z]{1} |
    [0-9]{11,11}
)$"""
ACTORS_MODEL = r"""^\s*(((?P<ragione_sociale1>[^0-9].+)\s+(?P<cf1>[a-zA-Z0-9]{11,}))|((?P<cf2>[a-zA-Z0-9]{11,})\s+(?P<ragione_sociale2>[^\(0-9].+)))\s*$"""
GROUP_ACTORS_MODEL = r"""^\s*(((?P<ragione_sociale1>[^0-9].+)\s+(?P<cf1>[a-zA-Z0-9]{11,}))|((?P<cf2>[a-zA-Z0-9]{11,})\s+(?P<ragione_sociale2>[^0-9].+)))\s*\(\s*(?P<ruolo>\d\d-[A-Z0-9-]+)\s*\)\s*$"""
DATES_MODEL = r"""^\s*(?P<start>\d\d\d\d-\d\d-\d\d)\s*[;\-/_ ]\s*(?P<end>\d\d\d\d-\d\d-\d\d)\s*$"""
IT_DATES_MODEL = r"""^\s*(?P<start_day>\d{1,2})[\-/](?P<start_month>\d{1,2})[\-/](?P<start_year>\d{4})\s*[;\-/_ ]\s*(?P<end_day>\d{1,2})[\-/](?P<end_month>\d{1,2})[\-/](?P<end_year>\d{4})\s*$""" 

SCELTA_CONTRAENTE_VOCABULARY = (
    "01-PROCEDURA APERTA",
    "02-PROCEDURA RISTRETTA",
    "03-PROCEDURA NEGOZIATA PREVIA PUBBLICAZIONE",
    "04-PROCEDURA NEGOZIATA SENZA PREVIA PUBBLICAZIONE",
    "05-DIALOGO COMPETITIVO",
    "06-PROCEDURA NEGOZIATA SENZA PREVIA INDIZIONE DI GARA (SETTORI SPECIALI)",
    "07-SISTEMA DINAMICO DI ACQUISIZIONE",
    "08-AFFIDAMENTO IN ECONOMIA - COTTIMO FIDUCIARIO",
    "14-PROCEDURA SELETTIVA EX ART 238 C.7, D.LGS. 163/2006",
    "17-AFFIDAMENTO DIRETTO EX ART. 5 DELLA LEGGE 381/91",
    "21-PROCEDURA RISTRETTA DERIVANTE DA AVVISI CON CUI SI INDICE LA GARA",
    "22-PROCEDURA NEGOZIATA CON PREVIA INDIZIONE DI GARA (SETTORI SPECIALI)",
    "23-AFFIDAMENTO DIRETTO",
    "24-AFFIDAMENTO DIRETTO A SOCIETA' IN HOUSE",
    "25-AFFIDAMENTO DIRETTO A SOCIETA' RAGGRUPPATE/CONSORZIATE O CONTROLLATE NELLE CONCESSIONI E NEI PARTENARIATI",
    "26-AFFIDAMENTO DIRETTO IN ADESIONE AD ACCORDO QUADRO/CONVENZIONE",
    "27-CONFRONTO COMPETITIVO IN ADESIONE AD ACCORDO QUADRO/CONVENZIONE",
    "28-PROCEDURA AI SENSI DEI REGOLAMENTI DEGLI ORGANI COSTITUZIONALI",
    "29-PROCEDURA RISTRETTA SEMPLIFICATA",
    "30-PROCEDURA DERIVANTE DA LEGGE REGIONALE",
    "31-AFFIDAMENTO DIRETTO PER VARIANTE SUPERIORE AL 20% DELL'IMPORTO CONTRATTUALE",
    "32-AFFIDAMENTO RISERVATO",
    "33-PROCEDURA NEGOZIATA PER AFFIDAMENTI SOTTO SOGLIA",
    "34-PROCEDURA ART.16 COMMA 2-BIS DPR 380/2001 PER OPERE URBANIZZAZIONE A SCOMPUTO PRIMARIE SOTTO SOGLIA COMUNITARIA",
    "35-PARTERNARIATO PER L'INNOVAZIONE",
    "36-AFFIDAMENTO DIRETTO PER LAVORI, SERVIZI O FORNITURE SUPPLEMENTARI",
    "37-PROCEDURA COMPETITIVA CON NEGOZIAZIONE",
    "38-PROCEDURA DISCIPLINATA DA REGOLAMENTO INTERNO PER SETTORI SPECIALI",
)

RUOLO_VOCABULARY = (
    "01-MANDANTE",
    "02-MANDATARIA",
    "03-ASSOCIATA",
    "04-CAPOGRUPPO",
    "05-CONSORZIATA",
)
